File: programmers/2020_KAKAO/블록이동하기/one.py
from collections import deque

def valid(r,c,N):
    return 0<=r<N and 0<=c<N 

def solution(board):
    N = len(board)
    visited = [[[0,0] for _ in range(N)] for _ in range(N)]
    q = deque()

    q.append([0,0,0])

    while q:
        r,c,p = q.popleft()
        t = visited[r][c][p]
        if p == 0:
            if r == N-1 and c+1 == N-1:
                return t

            new_r1, new_c1 = r-1, c
            new_r2, new_c2 = r-1, c+1
            if valid(new_r1,new_c1,N) and valid(new_r2,new_c2,N) and board[new_r1][new_c1]==0 and board[new_r2][new_c2]==0:
            # 1. 위로 회전1) 반시계방향
                if visited[r-1][c][1] == 0:
                    q.append([r-1, c, 1])
                    visited[r-1][c][1] = t+1
            # 2. 위로 회전2) 시계방향
                if visited[r-1][c+1][1] == 0:
                    q.append([r-1,c+1,1])
                    visited[r-1][c+1][1] = t+1
            # 3. 위로 평행 이동
                if visited[r-1][c][0] == 0:
                    q.append([r-1,c,0])
                    visited[r-1][c][0] = t+1

            new_r1, new_c1 = r+1, c
            new_r2, new_c2 = r+1, c+1
            if valid(new_r1,new_c1,N) and valid(new_r2,new_c2,N) and board[new_r1][new_c1]==0 and board[new_r2][new_c2]==0:
            # 4. 아래로 회전1) 시계방향
                if visited[r][c][1] == 0:
                    q.append([r, c, 1])
                    visited[r][c][1] = t+1
            # 5. 아래로 회전2) 반시계방향
                if visited[r][c+1][1] == 0:
                    q.append([r,c+1,1])
                    visited[r][c+1][1] = t+1
            # 6. 아래로 평행이동
                if visited[r+1][c][0] == 0:
                    q.append([r+1,c,0])
                    visited[r+1][c][0] = t+1
            
            # 7. 오른쪽으로 평행이동
            new_r, new_c = r, c+2
            if valid(new_r, new_c,N) and board[new_r][new_c] == 0:
                if visited[r][c+1][0] == 0:
                    q.append([r,c+1,0])
                    visited[r][c+1][0] = t+1

            # 8. 왼쪽으로 평행이동
            new_r, new_c = r, c-1
            if valid(new_r,new_c,N) and board[new_r][new_c] == 0:
                if visited[r][c-1][0] == 0:
                    q.append([r,c-1,0])
                    visited[r][c-1][0] = t+1
            
        else:
            if r+1 == N-1 and c == N-1:
                return t
            new_r1, new_c1 = r, c+1
            new_r2, new_c2 = r+1, c+1
            if valid(new_r1,new_c1,N) and valid(new_r2,new_c2,N) and board[new_r1][new_c1]==0 and board[new_r2][new_c2]==0:
            # 1. 오른쪽으로 회전1) 반시계방향
                if visited[r][c][0] == 0:
                    q.append([r, c, 0])
                    visited[r][c][0] = t+1
            # 2. 오른쪽으로 회전2) 시계방향
                if visited[r+1][c][0] == 0:
                    q.append([r+1,c,0])
                    visited[r+1][c][0] = t+1
            # 3. 오른쪽으로 평행이동
                if visited[r][c+1][1] == 0:
                    q.append([r,c+1,1])
                    visited[r][c+1][1] = t+1

            new_r1, new_c1 = r, c-1
            new_r2, new_c2 = r+1, c-1
            if valid(new_r1,new_c1,N) and valid(new_r2,new_c2,N) and board[new_r1][new_c1]==0 and board[new_r2][new_c2]==0:
            # 4. 왼쪽으로 회전1) 시계방향
                if visited[r][c-1][0] == 0:
                    q.append([r, c-1, 0])
                    visited[r][c-1][0] = t+1
            # 5. 왼쪽으로 회원2) 반시계방향
                if visited[r+1][c-1][0] == 0:
                    q.append([r+1,c-1,0])
                    visited[r+1][c-1][0] = t+1
            # 6. 왼쪽으로 평행이동
                if visited[r][c-1][1] == 0:
                    q.append([r,c-1,1])
                    visited[r][c-1][1] = t+1
            # 7. 위로 평행이동
            new_r, new_c = r-1, c
            if valid(new_r, new_c,N) and board[new_r][new_c] == 0:
                if visited[r-1][c][1] == 0:
                    q.append([r-1,c,1])
                    visited[r-1][c][1] = t+1
            # 8. 아래로 평행이동 
            new_r, new_c = r+2, c
            if valid(new_r,new_c,N) and board[new_r][new_c] == 0:
                if visited[r+1][c][1] == 0:
                    q.append([r+1,c,1])
                    visited[r+1][c][1] = t+1

    return 0


# board = [[0, 0, 0, 1, 1],[0, 0, 0, 1, 0],[0, 1, 0, 1, 1],[1, 1, 0, 0, 1],[0, 0, 0, 0, 0]]
board = [[0, 0, 0, 0, 0],[0, 0, 0, 0, 0],[0, 0, 0, 0, 0],[0, 0, 0, 0, 0],[0, 0, 0, 0, 0]]

print(solution(board))

File: programmers/2020_KAKAO/블록이동하기/one.js
function solution(board) {
  const N = board.length;
  const visited = Array.from(Array(N), () => Array(N).fill([0, 0]));
  const valid = (r, c, N) => 0 <= r && r < N && 0 <= c && c < N;
  const q = [];
  q.push([0, 0, 0]);
  while (q.length > 0) {
    let [r, c, p] = q.shift();
    let t = visited[r][c][p];

    // 가로로 누워있는 상태
    if (p === 0) {
      // 종료조건
      if (r === N - 1 && c + 1 === N - 1) {
        return t;
      }
      // 위쪽으로
      if (
        valid(r - 1, c, N) &&
        valid(r - 1, c + 1, N) &&
        board[r - 1][c] === 0 &&
        board[r - 1][c + 1] === 0
      ) {
        // 1. 반시계 방향
        if (visited[r - 1][c][1] === 0) {
          visited[r - 1][c][1] = t + 1;
          q.push([r - 1, c, 1]);
        }
        // 2. 시계 방향
        if (visited[r - 1][c + 1][1] === 0) {
          visited[r - 1][c + 1][1] = t + 1;
          q.push([r - 1, c + 1, 1]);
        }
        // 3. 평행이동
        if (visited[r - 1][c][0] === 0) {
          visited[r - 1][c][0] = t + 1;
          q.push([r - 1, c, 0]);
        }
        // 아래로
      }
      if (
        valid(r + 1, c, N) &&
        valid(r + 1, c + 1, N) &&
        board[r + 1][c] === 0 &&
        board[r + 1][c + 1] === 0
      ) {
        // 4. 시계방향
        if (visited[r][c][1] === 0) {
          visited[r][c][1] = t + 1;
          q.push([r, c, 1]);
        }
        // 5. 반시계방향
        if (visited[r][c + 1][1] === 0) {
          visited[r][c + 1][1] = t + 1;
          q.push([r, c + 1, 1]);
        }
        // 6. 평행이동
        if (visited[r + 1][c][0] === 0) {
          visited[r + 1][c][0] = t + 1;
          q.push([r + 1, c, 0]);
        }
      }
      // 7. 왼쪽으로
      if (valid(r, c - 1, N) && board[r][c - 1] === 0) {
        if (visited[r][c - 1][0] === 0) {
          visited[r][c - 1][0] = t + 1;
          q.push([r, c - 1, 0]);
        }
      }
      // 8. 오른쪽으로
      if (valid(r, c + 2, N) && board[r][c + 2] === 0) {
        if (visited[r][c + 1][0] === 0) {
          visited[r][c + 1][0] = t + 1;
          q.push([r, c + 1, 0]);
        }
      }

      // 세로로 서있는 상태
    } else {
      if (r + 1 === N - 1 && c === N - 1) {
        return t;
      }
      // 오른쪽으로
      if (
        valid(r, c + 1, N) &&
        valid(r + 1, c + 1, N) &&
        board[r][c + 1] === 0 &&
        board[r + 1][c + 1] === 0
      ) {
        // 1. 반시계방향
        if (visited[r][c][0] === 0) {
          visited[r][c][0] = t + 1;
          q.push([r, c, 0]);
        }
        // 2. 시계방향
        if (visited[r + 1][c][0] === 0) {
          visited[r + 1][c][0] = t + 1;
          q.push([r + 1][c][0]);
        }
        // 3. 평행이동
        if (visited[r][c + 1][1] === 0) {
          visited[r][c + 1][1] = t + 1;
          q.push([r, c + 1, 1]);
        }
      }
      // 왼쪽으로
      if (
        valid(r, c - 1, N) &&
        valid(r + 1, c - 1, N) &&
        board[r][c - 1] === 0 &&
        board[r + 1][c - 1] === 0
      ) {
        // 4. 시계방향
        if (visited[r][c - 1][0] === 0) {
          visited[r][c - 1][0] = t + 1;
          q.push([r, c - 1, 0]);
        }
        // 5. 반시계방향
        if (visited[r + 1][c - 1][0] === 0) {
          visited[r + 1][c - 1][0] = t + 1;
          q.push([r + 1, c - 1, 0]);
        }
        // 6. 평행이동
        if (visited[r][c - 1][0] === 0) {
          visited[r][c - 1][0] = t + 1;
          q.push([r, c - 1, 0]);
        }
      }
      // 7. 위로
      if (valid(r - 1, c, N) && board[r - 1][c] === 0) {
        if (visited[r - 1][c][1] === 0) {
          visited[r - 1][c][1] = t + 1;
          q.push([r - 1, c, 1]);
        }
      }

      // 8. 아래로
      if (valid(r + 2, c, N) && board[r + 2][c] === 0) {
        if (visited[r + 1][c][1] === 0) {
          visited[r + 1][c][1] = t + 1;
          q.push([r + 1, c, 1]);
        }
      }
    }
  }

  console.log(visited);

  return 0;
}

const board = [
  [0, 0, 0, 1, 1],
  [0, 0, 0, 1, 0],
  [0, 1, 0, 1, 1],
  [1, 1, 0, 0, 1],
  [0, 0, 0, 0, 0],
];
console.log(solution(board));
